test/grep.js

```diff
@@ -174,3 +174,225 @@ test('the pattern looks like an option', t => {
+      '11-line11\n' +
+      '12-line12\n' +
+      '--\n' +
+      '15:line15 test line\n\n'
+  );
+});
+
+test('-AB option, separated, same', t => {
+  const result = shell.grep('-AB', 3, 'test*', 'test/resources/grep/file3');
+  t.falsy(shell.error());
+  t.is(
+    result.toString(),
+    'line1\n' +
+      'line2 test line\n' +
+      'line3 test line\n' +
+      'line4\n' +
+      'line5\n' +
+      '--\n' +
+      'line1\n' +
+      'line2 test line\n' +
+      'line3 test line\n' +
+      'line4\n' +
+      'line5\n' +
+      'line6\n' +
+      '--\n' +
+      'line7\n' +
+      'line8\n' +
+      'line9\n' +
+      'line10 test line\n' +
+      'line11\n' +
+      'line12\n' +
+      'line13\n' +
+      '--\n' +
+      'line12\n' +
+      'line13\n' +
+      'line14\n' +
+      'line15 test line\n\n'
+  );
+});
+
+test('-AB option, separated, different', t => {
+  const result = shell.grep({ '-A': 2, '-B': 3 }, 'test*', 'test/resources/grep/file3');
+  t.falsy(shell.error());
+  t.is(
+    result.toString(),
+    'line1\n' +
+      'line2 test line\n' +
+      'line3 test line\n' +
+      'line4\n' +
+      '--\n' +
+      'line1\n' +
+      'line2 test line\n' +
+      'line3 test line\n' +
+      'line4\n' +
+      'line5\n' +
+      '--\n' +
+      'line7\n' +
+      'line8\n' +
+      'line9\n' +
+      'line10 test line\n' +
+      'line11\n' +
+      'line12\n' +
+      '--\n' +
+      'line12\n' +
+      'line13\n' +
+      'line14\n' +
+      'line15 test line\n\n'
+  );
+});
+
+test('-AB option, -n option, separated, same', t => {
+  const result = shell.grep('-nAB', 3, 'test*', 'test/resources/grep/file3');
+  t.falsy(shell.error());
+  t.is(
+    result.toString(),
+    '1-line1\n' +
+      '2:line2 test line\n' +
+      '3-line3 test line\n' +
+      '4-line4\n' +
+      '5-line5\n' +
+      '--\n' +
+      '1-line1\n' +
+      '2-line2 test line\n' +
+      '3:line3 test line\n' +
+      '4-line4\n' +
+      '5-line5\n' +
+      '6-line6\n' +
+      '--\n' +
+      '7-line7\n' +
+      '8-line8\n' +
+      '9-line9\n' +
+      '10:line10 test line\n' +
+      '11-line11\n' +
+      '12-line12\n' +
+      '13-line13\n' +
+      '--\n' +
+      '12-line12\n' +
+      '13-line13\n' +
+      '14-line14\n' +
+      '15:line15 test line\n\n'
+  );
+});
+
+test('-AB option, -n option, separated, different', t => {
+  const result = shell.grep(
+    { '-n': true, '-A': 2, '-B': 3 },
+    'test*', 'test/resources/grep/file3'
+  );
+  t.falsy(shell.error());
+  t.is(
+    result.toString(),
+    '1-line1\n' +
+      '2:line2 test line\n' +
+      '3-line3 test line\n' +
+      '4-line4\n' +
+      '--\n' +
+      '1-line1\n' +
+      '2-line2 test line\n' +
+      '3:line3 test line\n' +
+      '4-line4\n' +
+      '5-line5\n' +
+      '--\n' +
+      '7-line7\n' +
+      '8-line8\n' +
+      '9-line9\n' +
+      '10:line10 test line\n' +
+      '11-line11\n' +
+      '12-line12\n' +
+      '--\n' +
+      '12-line12\n' +
```

Comment:
This extra line separator doesn't match the behavior I see in unix grep. If the context regions overlap, then unix grep's behavior is to omit the line separator and treat it like one big region.

```
$ grep -A 2 -B 3 'test' test.txt
line1
line2 test line
line3 test line
line4
line5
--
line7
line8
line9
line10 test line
line11
line12
line13
line14
line15 test line
```

Reply:
I've fixed this, but I had to write an extra loop to solve the issue. That means it will be minimally slower but I'm not sure if there's a better option.